### PR TITLE
Prevent unneeded allocations of `Block` children enumerators.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
**Before:**
image

**After:**
image

#635

### CODE REVIEW

File: src/Microsoft.AspNet.Razor/Parser/SyntaxTree/Block.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ public Block(BlockBuilder source)
             source.Reset();
         }
 
-        protected Block(BlockType? type, IEnumerable<SyntaxTreeNode> contents, IParentChunkGenerator generator)
+        protected Block(BlockType? type, List<SyntaxTreeNode> contents, IParentChunkGenerator generator)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/Block.cs]: Should this be `IList<>` since it's on a publicish API?

Comment (rynowak) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/Block.cs]: Do we want to make these consistent and use `IReadOnlyList<>` here? (like the other constructor)

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/Block.cs]: `IList` doesn't implement `IReadOnlyList`.

File: src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ public BlockBuilder(Block original)
         }
 
         public BlockType? Type { get; set; }
-        public IList<SyntaxTreeNode> Children { get; private set; }
+        public List<SyntaxTreeNode> Children { get; private set; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs]: why?

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs [hunk 3]
```diff
@@ -302,8 +302,13 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
         private IEnumerable<string> GetAttributeNames(Block tagBlock)
         {
             // Need to calculate how many children we should take that represent the attributes.
-            var childrenOffset = IsPartialTag(tagBlock) ? 1 : 2;
-            var attributeChildren = tagBlock.Children.Skip(1).Take(tagBlock.Children.Count() - childrenOffset);
+            var childrenOffset = IsPartialTag(tagBlock) ? 0 : 1;
+            var childrenCount = tagBlock.Children.Count - childrenOffset;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs]: `childCount`?

File: src/Microsoft.AspNet.Razor/Parser/SyntaxTree/Block.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/Block.cs]: Oh shoot, this is `Block`. Yes `IReadOnlyList` works. Ignore my previous comment.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs [hunk 4]
```diff
@@ -302,8 +302,13 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
...
+            var attributeChildren = new List<SyntaxTreeNode>();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs]: Intialize with the capacity since you already know it

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs]: :+1:

File: src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs]: `List` : `IReadOnlyList`, `IList` does not.

Comment (rynowak) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs]: #BackingField

**Comment by rynowak** [comment]
:shipit: after fixing minor comments

File: src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs]: Needs to be exposed so `Block.cs` can consume it.

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs]: But `Block` doesn't need to change the collection does it? `IReadOnlyList` should work here w/ a `List` backing field.

File: test/Microsoft.AspNet.Razor.Test/Framework/BlockTypes.cs [hunk 5]
```diff
@@ -29,24 +29,19 @@ public StatementBlock(params SyntaxTreeNode[] children)
             : this(ParentChunkGenerator.Null, children)
         {
         }
-
-        public StatementBlock(IEnumerable<SyntaxTreeNode> children)
-            : this(ParentChunkGenerator.Null, children)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Framework/BlockTypes.cs]: Just curious: Were these constructor overloads ever used?

**Comment by dougbu** [comment]
:shipit: w/ @rynowak's carefully-considered suggestions fixed in a carefully-considered fashion.

File: test/Microsoft.AspNet.Razor.Test/Framework/BlockTypes.cs (hunk 5 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Framework/BlockTypes.cs]: Nope

File: src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/SyntaxTree/BlockBuilder.cs]: `Block` doesn't need to change it but this is a builder type. There'd be no way to add `Children` to `BlockBuilder`s unless I added an additional method to add children (not worth when its true type has the capability already).

**Pull request merged**